internal/model/modelindexer/indexer.go

```diff
@@ -167,9 +168,10 @@ func New(client elasticsearch.Client, cfg Config) (*Indexer, error) {
 func (i *Indexer) Close(ctx context.Context) error {
 	i.mu.Lock()
 	defer i.mu.Unlock()
-	if !i.closing {
-		i.closing = true
-		close(i.bulkItems)
+	select {
+	case <-i.closing:
+	default:
+		close(i.closing)
 		// Close i.closed when ctx is cancelled,
 		// unblocking any ongoing flush attempts.
```

Comment:
Instead of using `i.closed` will it be more readable to use errgroup context and propagate it? (In this case `i.closed` can be signaled by the errgroup context being canceled) and in `Close` we can just have our goroutine run as:

```go
i.g.Go(func() error {
    select {
    case <-done:
    case <-ctx.Done():
        return context.Canceled
    }
})
```

Reply:
That sounded like a great idea until I realised that we need to wait for errgroup.Wait to return in Close. The act of adding that Go call would cause Close to always block until the context passed to Close is cancelled.

Reply:
We could create a separate cancellable context for all flush calls though, which is cancelled by Close when the Close context is cancelled. That might be neater than what we have now.

Reply:
> until I realised that we need to wait for errgroup.Wait to return in Close

Ah, yes.

> We could create a separate cancellable context for all flush calls though, which is cancelled by Close when the Close context is cancelled. That might be neater than what we have now.

This looks good, I can follow it easily now. Thanks 😄